src/drivers/intel/ssp/ssp.c

```diff
@@ -67,22 +67,23 @@ static void ssp_empty_tx_fifo(struct dai *dai)
 /* empty SSP receive FIFO */
 static void ssp_empty_rx_fifo(struct dai *dai)
 {
+	struct timer *timer = timer_get();
+	uint64_t deadline = platform_timer_get(timer) +
+		clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) *
+		SSP_RX_FLUSH_TIMEOUT / 1000;
 	uint32_t sssr;
-	uint32_t entries;
-	uint32_t i;
 
-	sssr = ssp_read(dai, SSSR);
+	do {
+		while (ssp_read(dai, SSSR) & SSSR_RNE)
+			/* read to empty the fifo */
+			ssp_read(dai, SSDR);
+	} while ((ssp_read(dai, SSSR) & SSSR_BSY) &&
+		 platform_timer_get(timer) < deadline);
```

Comment:
Not following why we need a timeout here ? Isn't the FIFO 16 entries deep ?

Reply:
Just in case something refills the fifo too fast due to misconfiguration?

Reply:
The stream should not be active here, we are just clearing stale data from the FIFO.

Reply:
Broken driver might still leave it active in hardware even if in firmware it's not active. Better safe than sorry, I mean that's why the firmware even has panic handler in the first place.

Reply:
exactly as @paulstelian97 stated, we need to make sure SSSR_BSY is cleared and then  and SSSR_RNE is also cleared. The debugging shows that this SSSR_BSY is still active so there is following entries filled in to the RX FIFO.

And, this is actually follow the logic of the Linux driver, thanks @bkokoszx for spotting this:
https://github.com/torvalds/linux/blob/a1b8638ba1320e6684aa98233c15255eb803fac7/drivers/spi/spi-pxa2xx.c#L480

Reply:
@keyonjie the SSP rx link is disabled. At most there are FIFO_SIZE depth entries in the FIFO ? So this is an issue due to the readback FIFO being rate limited. Can we have a comment stating that FIFO readback is rate limited and the timeout is SSP_RX_FLUSH_TIMEOUT (btw do we read FIFO entries after we empty it ???)

Reply:
> @keyonjie the SSP rx link is disabled. At most there are FIFO_SIZE depth entries in the FIFO ? So this is an issue due to the readback FIFO being rate limited. Can we have a comment stating that FIFO readback is rate limited and the timeout is SSP_RX_FLUSH_TIMEOUT (btw do we read FIFO entries after we empty it ???)

@lgirdwood I think the result shows that after the first round of FIFO 'empty' with ssp_read(), SSSR_RNE is cleared so we thought it is 'empty', but the SSSR_BSY is still set which means the link is not fully stopped yet and there could be subsequent entries filled into the FIFO entries.

I would not assume that FIFO_SIZE is the maximum entries need to be emptied. I think it is not because of the FIFO readback is rate limited, opposite, the readback is very fast so emptying FIFO entries might need to take place for several rounds.

Reply:
@keyonjie ok, please explain this in a comment and we are good to go (otherwise anyone readying the code would easily get confused by this approach).